flooddrake/slope_limiter.py

```diff
@@ -24,74 +24,18 @@ def __init__(self, b, V, VCG):
             self.v, self.vu = split(self.V)
             self.v_cg, self.vu_cg = split(self.VCG)
 
-        # define the cg function to find maximum and minimum values for each vertex
-        self.v_max_cg = Function(self.v_cg)
-        self.v_min_cg = Function(self.v_cg)
-        self.vu_max_cg = Function(self.vu_cg)
-        self.vu_min_cg = Function(self.vu_cg)
-
-        if self.V.mesh().geometric_dimension() == 2:
-
-            self.vv_max_cg = Function(self.vv_cg)
-            self.vv_min_cg = Function(self.vv_cg)
-
-        # define bed and free surface
-        self.b_ = Function(self.v).project(self.b)
+        # define functions
         self.H = Function(self.v)
-        self.MU = Function(self.vu)
-
-        if self.V.mesh().geometric_dimension() == 2:
-
-            self.MV = Function(self.vv)
-
-        # cell average functions of depth and free surface
-        self.c = Function(FunctionSpace(self.v.mesh(), 'DG', 0))
-        self.cu = Function(FunctionSpace(self.v.mesh(), 'DG', 0))
         self.v_func = Function(self.v)
-        self.vu_func = Function(self.vu)
-
-        if self.V.mesh().geometric_dimension() == 2:
-
-            self.vv_func = Function(self.vv)
-            self.cv = Function(FunctionSpace(self.v.mesh(), 'DG', 0))
+        self.b_ = Function(self.v).project(self.b)
 
-        self.Project = Projector(self.v_func, self.c)
-        self.Projectu = Projector(self.vu_func, self.cu)
+        # define the vertex based limiter objects
+        self.SL = VertexBasedLimiter(self.v)
+        self.SLmu = VertexBasedLimiter(self.vu)
```

Comment:
You only need one limiter object, I think you can reuse it for multiple different fields, as long as the functionspaces match.